### PR TITLE
fix!: Avoid immediately jailing validators that are no longer opted-out

### PR DESCRIPTION
**Pull request by mpoke**
## Description

Closes: #1517 

This PR implements the solution from [this comment](https://github.com/cosmos/interchain-security/issues/1517#issuecomment-1857930734). Mainly, when a validator can no longer opt out, the `StartHeight` fields in its `SigningInfo` struct (in the slashing module) is set to the current height. As a result, the validator cannot be jailed for downtime for at least the sliding window for downtime slashing (`SignedBlocksWindow`). 

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [x] Added `!` to the type prefix if the change is [state-machine breaking](https://github.com/cosmos/interchain-security/blob/main/RELEASES.md#breaking-changes)
* [x] Confirmed this PR does not introduce changes requiring state migrations, OR migration code has been added to consumer and/or provider modules
* [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
* [x] Provided a link to the relevant issue or specification
* [x] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/00-intro.md)
* [x] Included the necessary unit and integration [tests](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#testing)
* [x] Added a changelog entry to `CHANGELOG.md`
* [x] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] Updated the relevant documentation or specification
* [x] Reviewed "Files changed" and left comments if necessary 
* [x] Confirmed all CI checks have passed
* [ ] If this PR is library API breaking, bump the go.mod version string of the repo, and follow through on a new major release

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` the type prefix if the change is state-machine breaking
* [ ] confirmed this PR does not introduce changes requiring state migrations, OR confirmed migration code has been added to consumer and/or provider modules
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

### CODE REVIEW

File: tests/integration/common.go [hunk 1]
```diff
@@ -597,20 +597,25 @@ func (suite *CCVTestSuite) GetConsumerEndpointClientAndConsState(
 
 // setupValidatorPowers delegates from the sender account to give all
 // validators on the provider chain 1000 power.
```
Comment (sainoe) [inline tests/integration/common.go:599]:
```suggestion
// validators on the provider chain the given voting powers.
```

File: tests/integration/soft_opt_out.go [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+
+	testCases := []struct {
+		name            string
+		downtimeFunc    func(*consumerKeeper.Keeper, *slashingkeeper.Keeper, []byte, int)
+		targetValidator int
+		expJailed       bool
+		expSlashPacket  bool
+	}{
+		{
+			"donwtime top 95%",
```
Comment (sainoe) [inline tests/integration/soft_opt_out.go:31]:
```suggestion
			"downtime top 95%",
```

**Review by p-offtermatt — COMMENTED** [review]
Curious about this: why reset the start height instead of the missed blocks counter?

afaict, the cleanest solution to me seems to be to, for every opt-out validator, set the missed blocks counter to zero in each block, before it is entered on chain.

This has a few advantages:
* No need for an extra field on validators (though it's still state machine breaking)
* Works out-of-the-box with existing downtime checkers (they will just not see the missed blocks)

I might be missing something though. I think the current solution is also fine, maybe not worth changing this for.

**Comment by mpoke** [comment]
> Curious about this: why reset the start height instead of the missed blocks counter?

@p-offtermatt I didn't want to mess with the downtime logic too much. If you set the missed blocks counter to zero, you risk having the counter negative and potentially never trigger a downtime event. See https://github.com/cosmos/cosmos-sdk/blob/v0.47.7/x/slashing/keeper/infractions.go#L41-L52. 

> afaict, the cleanest solution to me seems to be to, for every opt-out validator, set the missed blocks counter to zero in each block, before it is entered on chain.

That would be a lot of store writes and I don't see the advantage of doing it like this. 
 
> This has a few advantages:
> 
> * No need for an extra field on validators (though it's still state machine breaking)

Why is the extra field an issue? 

> * Works out-of-the-box with existing downtime checkers (they will just not see the missed blocks)

See my comment above. I think reseting the missed blocks counter would mess with the downtime logic.

**Comment by p-offtermatt** [comment]
Makes sense, thanks. I didn't take a close look at the missed-blocks logic in the slashing module (which is more intricate than I thought), and indeed I think your solution seems good under the constraints.

**Review by p-offtermatt — APPROVED** [review]

File: x/ccv/consumer/keeper/soft_opt_out.go [hunk 3]
```diff
@@ -73,3 +83,42 @@ func (k Keeper) GetSmallestNonOptOutPower(ctx sdk.Context) int64 {
...
 	return int64(binary.BigEndian.Uint64(bz))
 }
+
+func (k Keeper) UpdateSlashingSigningInfo(ctx sdk.Context) {
+	smallestNonOptOutPower := k.GetSmallestNonOptOutPower(ctx)
+
+	// Update SigningInfo for opted out validators
+	valset := k.GetAllCCValidator(ctx)
+	// sort validators by power ascending
+	sort.SliceStable(valset, func(i, j int) bool {
```
Comment (bermuell) [inline x/ccv/consumer/keeper/soft_opt_out.go:93]: valset should be ordered already by validator address (as claimed by SDK)
pls check if the re-ordering according to power is still needed as it comes with an overhead.

Comment (mpoke) [reply x/ccv/consumer/keeper/soft_opt_out.go]: Nice catch. See https://github.com/cosmos/interchain-security/pull/1549/commits/b00c874094af6b725fe1d0d97e5ee0a89e3e50eb

**Review by sainoe — APPROVED** [review]

**Pull request merged**